text/3742-mutate-non-mutable-locals.md

```diff
@@ -0,0 +1,249 @@
+- Feature Name: `mut_non_mut`
+- Start Date: 2024-12-13
+- RFC PR: [rust-lang/rfcs#3742](https://github.com/rust-lang/rfcs/pull/3742)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+the `mut` keyword has two meanings currently:
+ - When applied to a reference, it means the reference is not aliased
+ - When applied to a local binding, it means that the binding is allowed be re-assigned or borrowed using `&mut`
+
+The first of these is fundamental to Rust's safety guarantees and enables some optimizations.  The second,
+however, is more of a lint: the error message may help catch mistakes, but ignoring it doesn't violate
+type safety or subvert any of Rust's global correctness guarantees.
+
+This RFC proposes that the re-assignment or creation of a `&mut` to a local variable not
+marked with `mut` be changed from a hard error to a deny-by-default lint diagnostic (called
+`mut_non_mut` for the sake of discussion).
+
+# Motivation
+[motivation]: #motivation
+
+This idea has been discussed previously:
+
+ - [2014 Blog post](https://smallcultfollowing.com/babysteps/blog/2014/05/13/focusing-on-ownership/)
+ - [2021 IRLO thread](https://internals.rust-lang.org/t/lack-of-mut-in-bindings-as-a-deny-by-default-lint/15818)
+ - [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Idea.3A.20Downgrade.20.22immutable.20variable.E2.80.9C.20errors.20to.20lints)
+
+One benefit of this change is reduced friction: adding and removing of the need for `mut` on a given local
+can happen frequently when refactoring, similar to how variables and functions may change from used to unused
```

Comment:
I'd be curios if there's any data backing this statement. It's been a while since I've written significant amounts of Rust, but I don't think marking bindings as `mut` was ever an issue or particularly disruptive.

Reply:
I've not seen any hard data, but some people consider it a major inconvenience while others swear it's never an issue.  I think it depends on:
 - The way you think through problems (writing a bunch of code first then asking the compiler if it's correct vs "conversing" with the compiler while writing)
 - Tools in use (LSP/IDE vs old-fashioned edit/compile/debug loop)
 - Overall coding style (Static single-assignment and functional vs mutation-heavy and imperative)
 - Possibly also project type (library, cli, gui, server, or game)